incubator/hnc/pkg/reconcilers/hierarchy_config.go

```diff
@@ -186,59 +186,58 @@ func (r *HierarchyConfigReconciler) syncWithForest(log logr.Logger, nsInst *core
 	defer r.Forest.Unlock()
 	ns := r.Forest.Get(inst.ObjectMeta.Namespace)
 
-	// Clear locally-set conditions in the forest so we can set them to the latest.
+	// Clear locally-set conditions in the forest; we'll re-add them if they're still relevant. But
+	// first, record whether there were any critical ones since if this changes, we'll need to notify
+	// other namespaces.
 	hadCrit := ns.HasLocalCritCondition()
-	ns.ClearLocalCondition("")
+	ns.ClearLocalConditions()
 
-	r.syncOwner(log, inst, nsInst, ns)
+	// If this is a subnamespace, make sure .spec.parent is set correctly. Then sync the parent to the
+	// forest, and finally notify any relatives (including the parent) that might have been waiting
+	// for this namespace to be synced.
+	r.syncSubnamespaceParent(log, inst, nsInst, ns)
+	r.syncParent(log, inst, ns)
 	r.markExisting(log, ns)
 
-	r.syncParent(log, inst, ns)
-	inst.Status.Children = ns.ChildNames()
+	// Sync other spec and spec-like info
 	r.syncHNSes(log, ns, hnsnms)
 	ns.UpdateAllowCascadingDelete(inst.Spec.AllowCascadingDelete)
 
-	r.syncLabel(log, nsInst, ns)
-
-	// Sync all conditions. This should be placed at the end after all conditions are updated.
+	// Sync the status
+	inst.Status.Children = ns.ChildNames()
 	r.syncConditions(log, inst, ns, hadCrit)
+
+	// Sync the tree labels. This should go last since it can depend on the conditions.
+	r.syncLabel(log, nsInst, ns)
 }
 
-// syncOwner sets the parent to the owner and updates the HNSMissing condition if the HNS instance
-// is missing in the parent namespace according to the forest.  The namespace owner annotation is
-// the source of truth of the ownership (e.g. being a subnamespace), since modifying a namespace has
-// higher privilege than what HNC users can do.
-func (r *HierarchyConfigReconciler) syncOwner(log logr.Logger, inst *api.HierarchyConfiguration, nsInst *corev1.Namespace, ns *forest.Namespace) {
-	// Clear the HNSMissing condition if this is not a subnamespace or to
-	// reset it for the updated condition later.
-	ns.ClearConditionsByCode(log, api.HNSMissing)
```

Comment:
Seems `HNSMissing` will never be cleared once set if it's not cleared here. For example, when an hns was missing but a correct hns is somehow added again.

Reply:
Because HNSMissing is now a local condition (see line 240), it gets cleared with all the other local conditions at the beginning of `syncForest` (line 193).

Reply:
Thanks for the clarification!